src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -321,129 +573,101 @@ int MPL_gavl_tree_free(MPL_gavl_tree_t gavl_tree)
                 else
                     cur_node->right = NULL;
             }
-            if (gavl_tree_iptr->gavl_free_fn)
-                gavl_tree_iptr->gavl_free_fn((void *) dnode->val);
+
+            while (dnode->subnode_list) {
+                sub_node = dnode->subnode_list;
+                dnode->subnode_list = sub_node->next;
+                if (gavl_tree_iptr->gavl_free_fn)
+                    gavl_tree_iptr->gavl_free_fn((void *) sub_node->val);
+                MPL_free(sub_node);
+            }
             MPL_free(dnode);
         }
     }
     MPL_free(gavl_tree_iptr);
     return mpl_err;
 }
 
+/* MPL_gavl_tree_delete deletes all subnodes that intersect input key (addr, len).
+ * The delete algorithm is as follows:
+ * 1. binary search tree nodes. if find any node intersects input key, search the
+ *    subnode_list of that node and delete the first subnode that intersects input
+ *    key during search. If subnode_list is empty, delete that node from tree as well
+ * 2. repeat step 1 until all related subnodes are deleted */
 int MPL_gavl_tree_delete(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t len)
 {
```

Comment:
> The deletion logic seems problematic to me. What is the use case to delete intersected buffers? In both XPMEM and GPU IPC, should the caller always delete the node whose range exactly matches with the input?

In a real case, please consider network registration, if a user allocates a buffer [0, 100]. He first registers buffer [10, 20], then [0, 100]. Later, he frees buffer [0, 100], at this time, we should delete both [10, 20] and [0, 100] that's why we need to delete all intersecting buffer. This case can be extended to XPMEM (since XPMEM kernel module deals with buffer remapping issue, so we actually don't need to worry about it).

> * If a tree node [0, 100] contains [10,20]->[0,100]->[20,30] subnodes, and the input is [15,25], do you delete all three subnodes? I do not see a use case that follows this deletion logic. Can you please give me an example?

yes, I will delete all three subnodes.
In theory, user can pass any input to delete function, but the case you give won't happen in MPICH
delete function is called when free or the equivalent function is called by user.
Therefore, in MPICH, the input of the delete must be full buffer. In your case, your full buffer should be [0, 100] or superset of [0, 100], so [15, 25] is an impossible input.

> * It looks like that there is no way to distinguish which subnode should be used if subnode x is a subset of subnode y and the input matches both. E.g., subnodes are [10,20]->[0,100], input is [10,20]. You always return the first matched one. Then, why do we need maintain those subsets? Is it OK to delete [10,20] from the tree when [0,100] is inserted?

It is not correct design, please see my previous comments

Reply:
I think the concern that we incorrectly delete being used buffers (i.e., when inserting a superset buffer) is caused by two reasons:
1. Every time we delete a tree node, it also triggers the user's free function (e.g., to deregister the buffer from network). Why do we have to couple them together? If we consider the AVL tree is only a user space cache, I think implicit deletion (i.e., delete subset buffers at insert time) should be fine.

2. You concern is to delete the cached buffer on the sender(local) process. For actual use cases with network buffer registration, we need the MR key only for remote buffers (not local) to issue an RDMA request. Thus, we need cache remote buffers' <addr, key, size>. At search time on the local process, we search the tree node which matches with the remote buffer in the network transfer request. At deletion time, it does not trigger any network operation because the buffer is registered at remote process. Please see my explanation about dynamic window. I do not know any case that we need cache the locally registered buffer (we need store it somewhere in order to deregister it, but we don't have to use AVL for fast search).

The major reason why I am asking so many questions regarding the subnode design is because I feel that the code is overcomplicated. We really need reasons to justify whether it is really needed.